From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 22 Apr 2026 23:46:52 +0800
Subject: [PATCH] Leaf Cache random tick block status

Co-authored by: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
As part of: Leaf (https://github.com/Winds-Studio/Leaf/blob/86c87e1ba03840470ed21d6dc872463a9d579edf/leaf-server/minecraft-patches/features/0185-Cache-random-tick-block-status.patch)
Licensed under: MIT (https://github.com/Winds-Studio/Leaf/blob/86c87e1ba03840470ed21d6dc872463a9d579edf/licenses/MIT.txt)

diff --git a/net/minecraft/world/level/chunk/LevelChunkSection.java b/net/minecraft/world/level/chunk/LevelChunkSection.java
index d274149a977ed828b6bd02d88d2850521ede6c60..9fb22f1b633e2837b608b29a9f10f84719491957 100644
--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -19,6 +19,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
     private short fluidCount;
     private short tickingBlockCount;
     private short tickingFluidCount;
+    private boolean isRandomlyTickingBlocksStatus; // Leaf - Cache random tick block status
     public final PalettedContainer<BlockState> states;
     private PalettedContainer<Holder<Biome>> biomes; // CraftBukkit - read/write
 
@@ -53,6 +54,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
         this.tickingFluidCount = source.tickingFluidCount;
         this.states = source.states.copy();
         this.biomes = source.biomes.copy();
+        this.isRandomlyTickingBlocksStatus = this.tickingBlockCount > 0; // Leaf - Cache random tick block status
     }
 
     public LevelChunkSection(final PalettedContainer<BlockState> states, final PalettedContainer<Holder<Biome>> biomes) { // CraftBukkit - read/write
@@ -170,6 +172,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
         }
 
         this.updateBlockCallback(sectionX, sectionY, sectionZ, state, previous); // Paper - block counting
+        this.isRandomlyTickingBlocksStatus = this.tickingBlockCount > 0; // Leaf - Cache random tick block status
 
         return previous;
     }
@@ -187,7 +190,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
     }
 
     public boolean isRandomlyTickingBlocks() {
-        return this.tickingBlockCount > 0;
+        return isRandomlyTickingBlocksStatus; // Leaf - Cache random tick block status
     }
 
     public boolean isRandomlyTickingFluids() {
@@ -203,6 +206,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
         this.tickingFluidCount = (short)0;
         this.specialCollidingBlocks = (short)0;
         this.tickingBlocks.clear();
+        this.isRandomlyTickingBlocksStatus = false; // Leaf - Cache random tick block status
 
         if (this.maybeHas((final BlockState state) -> !state.isAir())) {
             final PalettedContainer.Data<BlockState> data = this.states.data;
@@ -236,6 +240,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
                 this.nonEmptyBlockCount += (short)paletteCount;
                 if (state.isRandomlyTicking()) {
                     this.tickingBlockCount += (short) paletteCount;
+                    this.isRandomlyTickingBlocksStatus = this.tickingBlockCount > 0; // Leaf - Cache random tick block status
                     final short[] raw = coordinates.elements();
                     final int rawLen = raw.length;
 
